<!DOCTYPE html>
<html>
  {% include 'header.html' %}
  {% include 'mapheader.html' %}
    <body>
      {% include 'banner.html' %}
      {% include 'menu.html' %}

      <section>
                <div class="intro">

                  <p>
                    COVID-19 PubSeq is a free and open online
                    bioinformatics public sequence resource with
                    on-the-fly analysis of
                    sequenced <a href="https://pubmed.ncbi.nlm.nih.gov/?term=sars-cov-2&sort=date">SARS-CoV-2</a>
                    samples that allows for a quick turnaround in
                    identification of new virus strains. PubSeq allows
                    anyone to upload sequence material in the form of
                    FASTA or FASTQ files with accompanying metadata
                    through the <a href="/upload">web interface</a>
                    or <a href="/apidoc">REST API</a>. For more
                    information see the <a href="/about">FAQ!</a>.
                  </p>
      <section id="map_view" class="map">
        <div id="mapid"></div>
      </section>

                  <a href="https://projectredcap.org/"><img class="img-right" src="static/image/REDCap.png" /></a>
                  <p>
                    Make your sequence
                    data <a href="https://en.wikipedia.org/wiki/FAIR_data">FAIR</a>. Upload
                    your SARS-CoV-2 sequence (FASTA or FASTQ formats)
                    with simple metadata (JSONLD) to
                    the <a href="/about">public sequence
                    resource</a>. The upload will trigger a recompute
                    with all available sequences into a Pangenome
                    available for
                    <a href="/download">download</a>!
                  </p>
                  <p>
                    Your uploaded sequence will automatically be
                    processed and incorporated into the public
                    pangenome with metadata using worklows from the
                    High Performance Open Biology Lab
                    defined <a href="https://github.com/hpobio-lab/viral-analysis/tree/master/cwl/pangenome-generate">here</a>. All
                    data is published under
                    a <a href="https://creativecommons.org/licenses/by/4.0/">Creative
                    Commons license</a> You can take the published
                    (GFA/RDF/FASTA) data and store it in a triple
                    store for further processing.
                    Clinical data can
                    be stored securely
                    at <a href="https://redcap-covid19.elixir-luxembourg.org/redcap/">REDCap</a>.
                  </p>
                  <a href="https://nanoporetech.com/"><img class="img-right" src="static/image/oxford-nanopore2.jpg" /></a>
                  <p>
                    Data can be uploaded from any sequencing platform
                    in FASTA format. We give special attention to
                    workflows for
                    the <a href="https://nanoporetech.com/">Oxford
                    Nanopore</a> - see
                    also <a href="https://pubmed.ncbi.nlm.nih.gov/?term=sars-cov-2+nanopore&sort=date">pubmed</a>
                    - because it offers an affordable platform that is
                    great for SARS-CoV-2 sequencing and
                    identification. In New Zealand the Oxford Nanopore
                    is used for all
                    <a href="https://www.esr.cri.nz/home/about-esr/media-releases/esr-performs-first-new-zealand-genome-sequencing-on-covid-19-sample/">tracing</a>.
                  </p>
                  <p>
                    Note that form fields contain
                    web <a href="https://en.wikipedia.org/wiki/Web_Ontology_Language">ontology
                    URI's</a>
                    for <a href="https://en.wikipedia.org/wiki/Wikipedia:Disambiguation">disambiguation</a>
                    and machine readable metadata. For examples of
                    use, see the <a href="/blog">BLOG</a>.
                  </p>
                </div>
        </section>

      {% include 'footer.html' %}


      <script type="text/javascript">
        let scriptRoot = {{ request.script_root|tojson|safe }}; // examples

        drawMap()

      </script>

   </body>
</html>
